API objects should be safe to ref() inside their constructors

geoffreygaren opened the pull request:
#### 966da8693f4b60deedaa3226f3641530ff7d5f0a
<pre>
API objects should be safe to ref() inside their constructors
<a href="https://bugs.webkit.org/show_bug.cgi?id=280278">https://bugs.webkit.org/show_bug.cgi?id=280278</a>
<a href="https://rdar.apple.com/136599625">rdar://136599625</a>

Reviewed by Chris Dumez and Ryosuke Niwa.

284023@main caused a crash because it instantiated a RefPtr inside an API
object&apos;s constructor.

API objects delegate refcounting to their ObjC wrappers. However, the wrapper
data member is not initialized until after construction is over, so RefPtr
inside a constructor is use of uninitialized memory.

Fix this by initializing the wrapper data member in our base class
constructor.

For now, we use a ThreadSpecific to avoid having to update all API
constructors and #ifdef them for COCOA vs non-COCOA. Longer term, it
would be nice to find a way to just pass the wrapper as a constructor
argument.

(Note: We can&apos;t just assign the wrapper before invoking the constructor
because that&apos;s Undefined Behavior, and an API test seems to demonstrate
that clang does in fact trip over it, at least some of the time.)

* Source/WebKit/Shared/API/APIObject.h:
(API::Object::constructInWrapper):

Canonical link: <a href="https://commits.webkit.org/284236@main">https://commits.webkit.org/284236@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/74c4e874805aa2e29c2279551ba6a3026bbdff08

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/68818 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/48212 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/21480 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/72889 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/19964 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/56009 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/19783 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/54848 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/13287 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/71884 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/44048 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/59418 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/35315 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/40713 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/18326 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/62662 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/17192 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/74584 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/12792 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/16443 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/62334 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/12831 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/59500 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/62372 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15277 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/10334 "Passed tests") | [⏳ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/macOS-Ventura-Release-WK2-Intel-Tests-EWS "Waiting to run tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/44010 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/45087 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/46282 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/44826 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebKit/Shared/API/APIObject.cpp:
```diff
@@ -33,9 +33,32 @@
+
+WTF::ThreadSpecific<CFTypeRef>& Object::wrapperUnderConstruction()
+{
+    static WTF::ThreadSpecific<CFTypeRef>* result;
```
Shouldn't need `WTF::` here and elsewhere in this PR.

cdumez replied:
Also, it would be nice to use `LazyNeverDestroyed` to avoid naked raw pointers like this in new code.

cdumez commented on Source/WebKit/Shared/API/APIObject.cpp:
```diff
@@ -33,9 +33,32 @@
+    std::call_once(
+        flag,
+        [] () {
+            result = new WTF::ThreadSpecific<CFTypeRef>();
```
Do we really need the `()` ?

cdumez commented on Source/WebKit/Shared/API/APIObject.cpp:
```diff
@@ -33,10 +33,34 @@
+{
+    static LazyNeverDestroyed<ThreadSpecific<HashMap<Object*, CFTypeRef>>> s_apiObjectsUnderConstruction;
+    static std::once_flag onceFlag;
+    std::call_once(onceFlag, [&] {
```
I doubt you need the `&`

geoffreygaren committed:
API objects should be safe to ref() inside their constructors
https://bugs.webkit.org/show_bug.cgi?id=280278
rdar://136599625

Reviewed by Chris Dumez and Ryosuke Niwa.

284023@main caused a crash because it instantiated a RefPtr inside an API
object's constructor.

API objects delegate refcounting to their ObjC wrappers. However, the wrapper
data member is not initialized until after construction is over, so RefPtr
inside a constructor is use of uninitialized memory.

Fix this by initializing the wrapper data member in our base class
constructor.

For now, we use a ThreadSpecific to avoid having to update all API
constructors and #ifdef them for COCOA vs non-COCOA. Longer term, it
would be nice to find a way to just pass the wrapper as a constructor
argument.

(Note: We can't just assign the wrapper before invoking the constructor
because that's Undefined Behavior, and an API test seems to demonstrate
that clang does in fact trip over it, at least some of the time.)

* Source/WebKit/Shared/API/APIObject.h:
(API::Object::constructInWrapper):

Canonical link: https://commits.webkit.org/284236@main
